### ISSUE TITLE
change zoom while a previous zoom is being animated does not work

### ISSUE THREAD

**Issue by javisantana** [Leaflet/Leaflet#1289]
given this code:

``` javascript

    var map = L.map('map').setView(L.latLng(50.5, 30.51), 0);
    // add layers
    // ...
    setTimeout(function() {
        map.setZoom(1);
        map.setZoom(2);
    }, 3000);
```

the the timeout callback is called and the zoon animation finish map.getZoom() returns 1 instead of 2.

the problem appears when you set the zoom while the animation didn't finish. The zoom attribute is changed but is overwritten when the animation finish (https://github.com/Leaflet/Leaflet/blob/master/src/map/anim/Map.ZoomAnimation.js#L150)

I dont really know what is the best approach to solve this, but raising the zoom process again work for us (*). We changed in
https://github.com/Leaflet/Leaflet/blob/master/src/map/anim/Map.ZoomAnimation.js#L19

``` javascript
if (this._animatingZoom) { return true; }
```

by

``` javascript
if (this._animatingZoom) { 
 this._animateToZoom = zoom;
 return true;
}
```

so in _resetView here ( https://github.com/Leaflet/Leaflet/blob/master/src/map/anim/Map.ZoomAnimation.js#L150 ) the zoom process is started again.

I dont know how leaflet works in deep so probably @mourner knows what is the best way to solve this.

I tested it on current master and 0.4.4 

(*) in cartodb we use an external model to store the zoom/center and other serializable map options so we need to manage and sync the center and zoom manually

**Comment by colllin** [comment]
@mourner Can you effectively stop CSS transitions by removing the transition class?  That is the attempt made in PR #3899 , which might address this issue.  I feel like some renaming / breaking up of functions might make that intention more clear in the PR.

**Comment by marcinn** [comment]
Hi.

There are several issues reported similar like this one. I've got same issue with Leaflet 1.4, but I don't want to report same issue as a new ticket. The problem is when I'm calling sequence like this:
```
map = L.Map('map', {center: A})
map.panTo(B, {animate: true, duration: 2});
map.setZoom(Z)
```

or sometimes

```
map.setView(B)
map.setZoom(Z)
```

The problem is that `setZoom()` is calling `setView`, which is stopping the current animation, getting map's center, and finally it is setting requested zoom level. When it occurs, the map's center is somewhere between state A and B (in rare cases of 2nd example, map's center does not change at all).

I think that `_stop()` called in such context should not only stop animation, but also should set immediately previously requested state, B here: 

``` 
map = L.Map('map', {center: A})
map.panTo(B, ...) // starts panning slowly
map.setZoom(Z)
// _stop() -- stops animation immediately
// set requested state, setCenter(A) here
// finally setView(getCenter(), Z)
```

After a quick look I see two possibilities:
* store destination state in `PosAnimation` instance and add `finish()` function, which will set the final state and call `stop()`
* set the destination inside `Map._onPanTransitionEnd` callback, or bind a new callback to `PosAnimation`'s end event, which will set the requested state.

Summing up - the expected behaviour in such case is to finish animation immediately (set the requested state),  and after that apply further changes (`setZoom` here).

Best wishes,
Marcin
